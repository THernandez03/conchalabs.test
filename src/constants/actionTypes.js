export const TOGGLE_MUSIC = 'TOGGLE_MUSIC';
export const PLAY_MUSIC = 'PLAY_MUSIC';
export const PAUSE_MUSIC = 'PAUSE_MUSIC';
export const SIGN_IN = 'SIGN_IN';
export const SIGN_OUT = 'SIGN_OUT';
