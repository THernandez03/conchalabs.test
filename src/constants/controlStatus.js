export const PLAYING = 'PLAYING';
export const PAUSED = 'PAUSED';
export const NOT_STARTED = 'NOT_STARTED';
