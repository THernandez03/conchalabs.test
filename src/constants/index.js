export * from './actionTypes';
export * from './controlStatus';
