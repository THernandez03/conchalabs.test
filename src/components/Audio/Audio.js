import { forwardRef, useEffect } from 'react';
import { useSelector } from 'react-redux';

import { useAudioDependencies } from '../../hooks';
import { PLAYING, PAUSED } from '../../constants/controlStatus';

export const Audio = forwardRef((props, ref) => {
  const { audioContext } = useAudioDependencies();
  const musicStatus = useSelector((state) => state.musicStatus);
  const selectedItem = useSelector((state) => state.selectedItem);

  const isPlaying = musicStatus === PLAYING;
  const isPaused = musicStatus === PAUSED;

  /* eslint-disable no-undefined */
  const src = selectedItem ? `sounds/${selectedItem}` : undefined;

  useEffect(() => {
    const toggleMusic = async () => {
      if (!src || !audioContext) return;

      if (isPlaying) {
        await audioContext.resume();
        ref.current.play();
      } else if (isPaused) {
        ref.current.pause();
      }
    };
    toggleMusic();
  }, [ref, audioContext, isPlaying, isPaused, src]);

  useEffect(() => {
    if (src && selectedItem && isPlaying) {
      ref.current.pause();
      ref.current.play();
    }
  }, [ref, selectedItem, isPlaying, src]);

  return (
    <audio ref={ref} loop src={src}>
      <track default kind="captions" />
    </audio>
  );
});

Audio.propTypes = {};

Audio.defaultProps = {};
