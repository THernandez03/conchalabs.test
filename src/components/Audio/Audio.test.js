import { createRef } from 'react';
import { render } from '@testing-library/react';
import { useSelector } from 'react-redux';

import { Audio } from './Audio';
import { PLAYING, PAUSED, NOT_STARTED } from '../../constants/controlStatus';

jest.mock('react-redux');
jest.mock('../../hooks', () => ({
  useAudioDependencies: () => ({
    audioContext: { resume: jest.fn() },
  }),
}));

describe('<Audio/>', () => {
  it('Should render an <audio> element', () => {
    useSelector
      .mockImplementationOnce(() => NOT_STARTED)
      .mockImplementationOnce(() => 'example.wav');

    const { container } = render(<Audio />);
    expect(container).toMatchSnapshot();
  });

  it('Should accept an empty "src"', () => {
    useSelector.mockImplementationOnce(() => NOT_STARTED);
    const { container } = render(<Audio />);
    expect(container.querySelector('audio').src).toBe('');
  });

  it('Should play the "src" when the musicStatus is PLAYING', () => {
    useSelector
      .mockImplementationOnce(() => PLAYING)
      .mockImplementationOnce(() => 'example.wav');

    const playFn = jest.fn();
    const pauseFn = jest.fn();

    window.HTMLMediaElement.prototype.play = playFn;
    window.HTMLMediaElement.prototype.pause = pauseFn;

    render(<Audio ref={createRef()} />);
    expect(playFn).toHaveBeenCalled();
    expect(pauseFn).toHaveBeenCalledTimes(1);
  });

  it('Should pause the "src" when the musicStatus is PAUSED', () => {
    useSelector
      .mockImplementationOnce(() => PAUSED)
      .mockImplementationOnce(() => 'example.wav');

    const playFn = jest.fn();
    const pauseFn = jest.fn();

    window.HTMLMediaElement.prototype.play = playFn;
    window.HTMLMediaElement.prototype.pause = pauseFn;

    render(<Audio ref={createRef()} />);
    expect(playFn).not.toHaveBeenCalled();
    expect(pauseFn).toHaveBeenCalled();
  });
});
