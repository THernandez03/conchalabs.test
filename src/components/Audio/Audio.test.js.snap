// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Audio/> Should render an <audio> element 1`] = `
<div>
  <audio
    loop=""
    src="sounds/example.wav"
  >
    <track
      default=""
      kind="captions"
    />
  </audio>
</div>
`;
